### PR TITLE
Add batched quantities replacing for model data

### PR DESCRIPTION
**Pull request by flferretti**
This PR allows to replace quantities from vectorized `JaxSimModelData` instances.

----
📚 Documentation preview 📚: https://jaxsim--387.org.readthedocs.build//387/

### CODE REVIEW

File: src/jaxsim/api/integrators.py [hunk 1]
```diff
@@ -58,7 +58,7 @@ def semi_implicit_euler_integration(
         W_p_B = data.base_position + dt * W_ṗ_B
         W_Q_B = data.base_orientation + dt * W_Q̇_B
 
-        base_quaternion_norm = jaxsim.math.safe_norm(W_Q_B)
+        base_quaternion_norm = jaxsim.math.safe_norm(W_Q_B, axis=W_Q_B.ndim - 1)
```
Comment (younik) [inline src/jaxsim/api/integrators.py:61]: (see other comments)

File: src/jaxsim/api/data.py [hunk 2]
```diff
@@ -421,44 +432,69 @@ def replace(
...
+            ),
+            base_angular_velocity_inertial=jnp.broadcast_to(
+                base_angular_velocity_inertial, (batch_size, 3)
+            ),
         )
 
+        # Adjust the output shapes.
+        if batch_size == 1:
+            link_transforms = link_transforms.reshape(link_transforms.shape[1:])
+            link_velocities = link_velocities.reshape(link_velocities.shape[1:])
```
Comment (younik) [inline src/jaxsim/api/data.py:490]: I am surprised this works with jit(?)

File: src/jaxsim/api/data.py [hunk 3]
```diff
@@ -413,6 +416,14 @@ def replace(
         if base_position is None:
             base_position = self.base_position
 
+        # Normalize the quaternion to avoid numerical issues.
+        base_quaternion_norm = jaxsim.math.safe_norm(
+            base_quaternion, axis=base_quaternion.ndim - 1
```
Comment (younik) [inline src/jaxsim/api/data.py:421]: (see other comments)

File: src/jaxsim/api/data.py [hunk 4]
```diff
@@ -363,7 +360,7 @@ def reset_base_quaternion(self, base_quaternion: jtp.VectorLike) -> Self:
 
         W_Q_B = jnp.array(base_quaternion, dtype=float)
 
-        norm = jaxsim.math.safe_norm(W_Q_B)
+        norm = jaxsim.math.safe_norm(W_Q_B, axis=W_Q_B.ndim - 1)
```
Comment (younik) [inline src/jaxsim/api/data.py:363]: see other comments

File: src/jaxsim/api/data.py [hunk 5]
```diff
@@ -265,14 +265,14 @@ def base_orientation(self) -> jtp.Matrix:
...
-        W_Q_B = self.base_quaternion.squeeze()
+        W_Q_B = self.base_quaternion
 
         # Always normalize the quaternion to avoid numerical issues.
         # If the active scheme does not integrate the quaternion on its manifold,
         # we introduce a Baumgarte stabilization to let the quaternion converge to
         # a unit quaternion. In this case, it is not guaranteed that the quaternion
         # stored in the state is a unit quaternion.
-        norm = jaxsim.math.safe_norm(W_Q_B)
+        norm = jaxsim.math.safe_norm(W_Q_B, axis=W_Q_B.ndim - 1)
```
Comment (younik) [inline src/jaxsim/api/data.py:275]: see other comments

File: src/jaxsim/api/common.py [hunk 6]
```diff
@@ -188,47 +188,45 @@ def other_representation_to_inertial(
...
                 else:
-                    W_Xf_O = Adjoint.from_transform(transform=W_H_O, inverse=True).T
-                    W_array = W_Xf_O @ O_array
+                    W_Xf_O = Adjoint.from_transform(transform=W_H_O, inverse=True)
+                    W_array = jnp.einsum(
+                        "bij,bj->bi", W_Xf_O.transpose(0, 2, 1), O_array
+                    )
 
-                return W_array
+                return W_array.reshape(array.shape[:-1] + (6,))
```
Comment (younik) [inline src/jaxsim/api/common.py:210]: Not a big deal, but it should be possible to avoid reshaping by using `...` instead of b in einsum

> To enable and control broadcasting, use an ellipsis. Default NumPy-style broadcasting is done by adding an ellipsis to the left of each term, like np.einsum('...ii->...i', a). np.einsum('...i->...', a) is like [np.sum(a, axis=-1)](https://numpy.org/doc/2.1/reference/generated/numpy.sum.html#numpy.sum) for array a of any shape. To take the trace along the first and last axes, you can do np.einsum('i...i', a), or to do a matrix-matrix product with the left-most indices instead of rightmost, one can do np.einsum('ij...,jk...->ik...', a, b).

File: src/jaxsim/api/common.py [hunk 7]
```diff
@@ -121,45 +121,45 @@ def inertial_to_other_representation(
...
                 else:
-                    OW_Xf_W = Adjoint.from_transform(transform=W_H_OW).T
-                    OW_array = OW_Xf_W @ W_array
+                    OW_Xf_W = Adjoint.from_transform(transform=W_H_OW)
+                    OW_array = jnp.einsum(
+                        "bij,bj->bi", OW_Xf_W.transpose(0, 2, 1), W_array
+                    )
 
-                return OW_array
+                return OW_array.reshape(array.shape[:-1] + (6,))
```
Comment (younik) [inline src/jaxsim/api/common.py:162]: (see other comment)

**Review by younik — COMMENTED** [review]
It looks good! I just have few comments :)

**Review by xela-95 — CHANGES_REQUESTED** [review]
Great work @flferretti !! Just minor comments

File: src/jaxsim/api/common.py (hunk 7 above)
Comment (flferretti) [reply src/jaxsim/api/common.py]: Addressed in https://github.com/ami-iit/jaxsim/pull/387/commits/3565706462f9778f4d09050bc7026ff832045f2b, thanks!

File: src/jaxsim/api/common.py (hunk 6 above)
Comment (flferretti) [reply src/jaxsim/api/common.py]: Addressed in https://github.com/ami-iit/jaxsim/pull/387/commits/3565706462f9778f4d09050bc7026ff832045f2b, thanks!

File: src/jaxsim/api/data.py (hunk 5 above)
Comment (flferretti) [reply src/jaxsim/api/data.py]: Addressed in https://github.com/ami-iit/jaxsim/pull/387/commits/3565706462f9778f4d09050bc7026ff832045f2b, thanks!

File: src/jaxsim/api/data.py (hunk 3 above)
Comment (flferretti) [reply src/jaxsim/api/data.py]: Addressed in https://github.com/ami-iit/jaxsim/pull/387/commits/3565706462f9778f4d09050bc7026ff832045f2b, thanks!

File: src/jaxsim/api/integrators.py (hunk 1 above)
Comment (flferretti) [reply src/jaxsim/api/integrators.py]: Addressed in https://github.com/ami-iit/jaxsim/pull/387/commits/3565706462f9778f4d09050bc7026ff832045f2b, thanks!

File: src/jaxsim/api/integrators.py [hunk 8]
```diff
@@ -104,7 +104,9 @@ def f(x) -> dict[str, jtp.Matrix]:
                 joint_torques=joint_torques,
             )
 
-    base_quaternion_norm = jaxsim.math.safe_norm(data._base_quaternion)
+    base_quaternion_norm = jaxsim.math.safe_norm(
+        data._base_quaternion, axis=data._base_quaternion.ndim - 1
```
[the remark this answers is not in the document]
Comment (flferretti) [reply src/jaxsim/api/integrators.py]: Addressed in https://github.com/ami-iit/jaxsim/pull/387/commits/3565706462f9778f4d09050bc7026ff832045f2b, thanks!

File: src/jaxsim/math/utils.py [hunk 9]
```diff
@@ -28,4 +28,4 @@ def safe_norm(array: jtp.ArrayLike, axis=None) -> jtp.Array:
 
     # Use `jnp.where` to set the norm to 0.0 where the input array was all zeros.
     # This usage supports potential batch processing for future scalability.
-    return jnp.where(is_zero, 0.0, norm)
+    return jnp.where(is_zero, 0.0, norm).reshape(array.shape[:-1] + (1,))
```
[the remark this answers is not in the document]
Comment (flferretti) [reply src/jaxsim/math/utils.py]: Addressed in https://github.com/ami-iit/jaxsim/pull/387/commits/3565706462f9778f4d09050bc7026ff832045f2b, thanks!

File: src/jaxsim/rbda/contacts/relaxed_rigid.py [hunk 10]
```diff
@@ -703,7 +699,7 @@ def _hunt_crossley_contact_model(
 
         # Compute the direction of the tangential force.
         # To prevent dividing by zero, we use a switch statement.
-        norm = jaxsim.math.safe_norm(f_tangential)
+        norm = jaxsim.math.safe_norm(f_tangential, axis=f_tangential.ndim - 1)
```
[the remark this answers is not in the document]
Comment (flferretti) [reply src/jaxsim/rbda/contacts/relaxed_rigid.py]: Addressed in https://github.com/ami-iit/jaxsim/pull/387/commits/3565706462f9778f4d09050bc7026ff832045f2b, thanks!

File: src/jaxsim/api/data.py (hunk 4 above)
Comment (flferretti) [reply src/jaxsim/api/data.py]: Addressed in https://github.com/ami-iit/jaxsim/pull/387/commits/3565706462f9778f4d09050bc7026ff832045f2b, thanks!

File: src/jaxsim/api/common.py [hunk 11]
```diff
@@ -188,47 +188,45 @@ def other_representation_to_inertial(
...
                 if not is_force:
-                    W_Xv_O: jtp.Array = Adjoint.from_transform(W_H_O)
-                    W_array = W_Xv_O @ O_array
+                    W_Xv_O = Adjoint.from_transform(W_H_O)
+                    W_array = jnp.einsum("bij,bj->bi", W_Xv_O, O_array)
 
                 else:
-                    W_Xf_O = Adjoint.from_transform(transform=W_H_O, inverse=True).T
-                    W_array = W_Xf_O @ O_array
+                    W_Xf_O = Adjoint.from_transform(transform=W_H_O, inverse=True)
```
[the remark this answers is not in the document]
Comment (flferretti) [reply src/jaxsim/api/common.py]: Thanks! Addressed in https://github.com/ami-iit/jaxsim/pull/387/commits/fe34e75e364a133bf09395d5858499229cc1825b

File: src/jaxsim/api/common.py [hunk 12]
```diff
@@ -188,47 +188,45 @@ def other_representation_to_inertial(
...
                 if not is_force:
-                    W_Xv_BW: jtp.Array = Adjoint.from_transform(W_H_OW)
-                    W_array = W_Xv_BW @ BW_array
+                    W_Xv_BW = Adjoint.from_transform(W_H_OW)
+                    W_array = jnp.einsum("bij,bj->bi", W_Xv_BW, O_array)
 
                 else:
-                    W_Xf_BW = Adjoint.from_transform(transform=W_H_OW, inverse=True).T
-                    W_array = W_Xf_BW @ BW_array
+                    W_Xf_BW = Adjoint.from_transform(transform=W_H_OW, inverse=True)
```
[the remark this answers is not in the document]
Comment (flferretti) [reply src/jaxsim/api/common.py]: Thanks! Addressed in https://github.com/ami-iit/jaxsim/pull/387/commits/fe34e75e364a133bf09395d5858499229cc1825b

File: src/jaxsim/api/common.py [hunk 13]
```diff
@@ -121,45 +121,45 @@ def inertial_to_other_representation(
...
 
                 if not is_force:
                     OW_Xv_W = Adjoint.from_transform(transform=W_H_OW, inverse=True)
-                    OW_array = OW_Xv_W @ W_array
+                    OW_array = jnp.einsum("bij,bj->bi", OW_Xv_W, W_array)
 
                 else:
-                    OW_Xf_W = Adjoint.from_transform(transform=W_H_OW).T
-                    OW_array = OW_Xf_W @ W_array
+                    OW_Xf_W = Adjoint.from_transform(transform=W_H_OW)
```
[the remark this answers is not in the document]
Comment (flferretti) [reply src/jaxsim/api/common.py]: Thanks! Addressed in https://github.com/ami-iit/jaxsim/pull/387/commits/fe34e75e364a133bf09395d5858499229cc1825b

File: src/jaxsim/api/common.py (hunk 6 above)
Comment (flferretti) [reply src/jaxsim/api/common.py]: Thanks, addressed in https://github.com/ami-iit/jaxsim/pull/387/commits/1b37c461679802de43637767a0d3f619caf54674

File: src/jaxsim/api/common.py (hunk 7 above)
Comment (flferretti) [reply src/jaxsim/api/common.py]: Thanks, addressed in https://github.com/ami-iit/jaxsim/pull/387/commits/1b37c461679802de43637767a0d3f619caf54674

**Review by xela-95 — APPROVED** [review]

File: src/jaxsim/api/data.py (hunk 2 above)
Comment (younik) [reply src/jaxsim/api/data.py:490]: I think the latter is better!
Sorry for the lag

**Review by younik — APPROVED** [review]
Looks great, thanks!

File: src/jaxsim/api/data.py (hunk 2 above)
Comment (flferretti) [reply src/jaxsim/api/data.py:490]: I'm sorry, I missed this. I'll add that to #160
